piff/knn_interp.py

```diff
@@ -0,0 +1,373 @@
+            # if no y, then interpolate
+            y = numpy.array([self.knn[key].predict(X) for key in self.attr_target]).T
+        if logger:
+            logger.debug('Regression shape: %s', y.shape)
+        # add misalignment shape (n_targets, 3)
+        # X is (n_samples, 2)
+        # y is (n_samples, n_targets)
+        y = y + self.misalignment[numpy.newaxis, :, 0] \
+                + X[:, 1, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 1] \
+                + X[:, 0, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 2]
+
+        return y
+
+    def writeSolution(self, fits, extname):
+        """Write the solution to a FITS binary table.
+
+        Save the knn params and the X and y arrays
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+
+        dtypes = [('X', self.X.dtype, self.X.shape), ('Y', self.y.dtype, self.y.shape),
+                  ('XPIXEL', self.Xpixel.dtype, self.Xpixel.shape),
+                  ('ATTR_TARGET', self.attr_target.dtype, self.attr_target.shape),
+                  ('ATTR_INTERP', self.attr_interp.dtype, self.attr_interp.shape),
+                  ('MISALIGNMENT', self.misalignment.dtype, self.misalignment.shape)]
+        data = numpy.empty(1, dtype=dtypes)
+        # assign
+        data['X'] = self.X
+        data['Y'] = self.y
+        data['XPIXEL'] = self.Xpixel
+        data['ATTR_TARGET'] = self.attr_target
+        data['ATTR_INTERP'] = self.attr_interp
+        data['MISALIGNMENT'] = self.misalignment
+
+        # put the knn params in the header?
+        header = self.kwargs
+
+        # write to fits
+        fits.write_table(data, extname=extname, header=header)
+
+    def readSolution(self, fits, extname):
+        """Read the solution from a FITS binary table.
+
+        The extension should contain the same values as are saved
+        in the writeSolution method.
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+        # header = fits[extname].read_header()
+        data = fits[extname].read()
+
+        # kwargs come from base class read()
+        # self.kwargs = header
+
+        # attr_target and attr_interp assigned in build
+        self.build(data['ATTR_INTERP'][0], data['ATTR_TARGET'][0])
+        # self.X and self.y assigned in _fit
+        self._fit(data['X'][0], data['Y'][0])
+        self.misalign_wavefront(data['MISALIGNMENT'][0])
+
+        # other attributes
+        self.Xpixel = data['XPIXEL'][0]
+
+
+# pretty sure the correct way to do this is in the WCS framework, but let's
+# move forward first:
+def pixel_to_focal(stardata):
+    """Take stardata and add focal plane position to properties
+
+    :param stardata:    The stardata with property 'ccdnum'
+
+    :returns stardata:  New stardata with updated properties
+    """
```

Comment:
Chris etal,

Folks,

I had assumed that u,v were the “sky” version of the focal plane coordinates - ie. not RA,dec — is that correct?

Then if these are focal plane coordinates then I am also curious to know exactly how are they determined?

Chris is right to point out this issue with decenter - and I think it does suggest a reason to stay in focal plane coordinates

Aaron


On Jul 4, 2016, at 7:50 PM, Chris Davis <notifications@github.com<mailto:notifications@github.com>> wrote:


In piff/knn_interp.py<https://github.com/rmjarvis/Piff/pull/24#discussion_r69485366>:

> +        self._fit(data['X'][0], data['Y'][0])
> +        self.misalign_wavefront(data['MISALIGNMENT'][0])
> +
> +        # other attributes
> +        self.Xpixel = data['XPIXEL'][0]
> +
> +
> +# pretty sure the correct way to do this is in the WCS framework, but let's
> +# move forward first:
> +def pixel_to_focal(stardata):
> +    """Take stardata and add focal plane position to properties
> +
> +    :param stardata:    The stardata with property 'ccdnum'
> +
> +    :returns stardata:  New stardata with updated properties
> +    """


Sorry this probably demonstrates my ignorance re: wcs coordinates. Are the u,v coordinates based on the sky, or fixed to the camera? (ie do they change with exposure?)

I included this focal_x stuff because the wavefront model is defined in the coordinate system of the camera. There are two pieces to this. The first is that the reference wavefront (which uses the kNN interpolation to tell you what the defocus is at some point on the camera) is currently defined in terms of location on the camera, though it could be changed to be defined in terms of pixel coordinates on a given ccd (so e.g. that it loads up the kNN interpolation for ccd and zernike coefficient; currently it just loads up the kNN interpolation for zernike coefficient). The second piece is that misalignments of the camera effect the wavefront in a focal-plane dependent manner. Thus, if I were to add some decentering to the camera, there is a change in the astigmatisms that depends on where you are located on the camera.

I also wasn't sure how to implement this in the WCS, but I figured I should move forward with this and have you tell me how I go about that :)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/rmjarvis/Piff/pull/24/files/02a1fef3b765a4d95d314f43082b4a18f76706d2#r69485366>, or mute the thread<https://github.com/notifications/unsubscribe/AEEa8yDmOoxZUxC1yS59DekzLoi4SRJNks5qSVXrgaJpZM4I6RGE>.

_________________________________________________________________________
 Prof. Aaron Roodman
 Chair Dept. of Particle Physics & Astrophysics
 SLAC National Accelerator Laboratory
 Stanford University


 SLAC National Accelerator Laboratory        E-mail: roodman@slac.stanford.edu<mailto:roodman@slac.stanford.edu>
 2575 Sand Hill Rd.                                           Phone:  650-926-2705
 MS 29                                                                  Fax:  650-926-2657
 Menlo Park, CA 94025                                    URL: http://www.slac.stanford.edu/~roodman
_________________________________________________________________________